src/test_main.py

```diff
@@ -59,35 +60,231 @@ def test_generate_state_space_for_N_eq_1_and_k_eq_3():
+            source=source, target=target, fitness_function=symbolic_fitness_function
+        )
+        is None
+    )
+
+
+def test_generate_transition_matrix_for_trivial_fitness_function():
+    """"""
+
+    def trivial_fitness_function(state):
+        return np.array([1 for _ in state])
+
+    state_space = np.array(
+        [
+            (0, 0, 1),
+            (0, 1, 0),
+            (1, 0, 0),
+            (0, 1, 1),
+            (1, 0, 1),
+            (1, 1, 0),
+            (0, 0, 0),
+            (1, 1, 1),
+        ]
+    )
+    expected_transition_matrix = np.array(
+        [
+            [5 / 9, 0, 0, 1 / 9, 1 / 9, 0, 2 / 9, 0],
+            [0, 5 / 9, 0, 1 / 9, 0, 1 / 9, 2 / 9, 0],
+            [0, 0, 5 / 9, 0, 1 / 9, 1 / 9, 2 / 9, 0],
+            [1 / 9, 1 / 9, 0, 5 / 9, 0, 0, 0, 2 / 9],
+            [1 / 9, 0, 1 / 9, 0, 5 / 9, 0, 0, 2 / 9],
+            [0, 1 / 9, 1 / 9, 0, 0, 5 / 9, 0, 2 / 9],
+            [0, 0, 0, 0, 0, 0, 1, 0],
+            [0, 0, 0, 0, 0, 0, 0, 1],
+        ]
+    )
+    assert np.array_equal(
+        main.generate_transition_matrix(
+            state_space=state_space, fitness_function=trivial_fitness_function
+        ),
+        expected_transition_matrix,
+    )
 
-    Expected_Transition_Matrix = np.array([[1, 0, 0, 1, 1, 0, 1, 0],
-[0, 1, 0, 1, 0, 1, 1, 0],
-[0, 0, 1, 0, 1, 1, 1, 0],
-[1, 1, 0, 1, 0, 0, 0, 1],
-[1, 0, 1, 0, 1, 0, 0, 1],
-[0, 1, 1, 0, 0, 1, 0, 1],
-[1, 1, 1, 0, 0, 0, 1, 0],
-[0, 0, 0, 1, 1, 1, 0, 1]])
 
-    Obtained_Transition_Matrix = main.gen_transition_matrix(S)
-    assert (Obtained_Transition_Matrix == Expected_Transition_Matrix).all()
+def test_generate_transition_matrix_for_ordered_fitness_function():
+    """TODO"""
 
+    def ordered_fitness_function(
+        state,
+    ):  # From left to right, counts how many of the same type have been read, and adds the position % 2 to give fitness
+        fitness = np.array([0 for _ in state])
+        zero_encountered = 0
+        one_encountered = 0
+        for position, value in enumerate(state):
+            if value == 0:
+                zero_encountered += 1
+                fitness[position] = zero_encountered + (position % 2)
+            else:
+                one_encountered += 1
+                fitness[position] = one_encountered + (position % 2)
+        return fitness
 
-test_get_transition_matrix_for_given_state_space()
\ No newline at end of file
+    state_space = np.array(
+        [
+            (0, 0, 1),
+            (0, 1, 0),
+            (1, 0, 0),
+            (0, 1, 1),
+            (1, 0, 1),
+            (1, 1, 0),
+            (0, 0, 0),
+            (1, 1, 1),
+        ]
+    )
+    expected_transition_matrix = np.array(
+        [
+            [3 / 5, 0, 0, 1 / 15, 1 / 15, 0, 4 / 15, 0],
+            [0, 8 / 15, 0, 2 / 15, 0, 2 / 15, 1 / 5, 0],
+            [0, 0, 9 / 15, 0, 1 / 15, 1 / 15, 4 / 15, 0],
+            [1 / 15, 1 / 15, 0, 9 / 15, 0, 0, 0, 4 / 15],
```

Comment:
This row sums to 16/15 I believe?


After we figure out what's wrong, add an assert that the row sum is 1 for all the obtained ones.